### PR TITLE
Implemented Discord's Rich Presence

### PR DESCRIPTION
**Pull request by connorjan**
Still some stuff to do, but it should be in a stable enough state for playtesters.

### CODE REVIEW

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+#include "fmtstr.h"
+#include "mom_system_discord.h"
+#include "mom_shareddefs.h"
+#include <time.h>
+
+#include "tier0/memdbgon.h"
+
+// MOM_TODO: Change this to the official Discord App ID (currently it is under Connor's test account)
+const char* CMomentumDiscord::s_pDiscordAppID = "452297955718987776";
+const char* CMomentumDiscord::s_pSteamAppID = "669270";
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: These should be `#define`s instead

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+    m_szDiscordLargeImageKey[0] = '\0';
+    m_szDiscordLargeImageText[0] = '\0';
+    m_szDiscordSmallImageKey[0] = '\0';
+    m_szDiscordSmallImageText[0] = '\0';
+    m_szDiscordPartyId[0] = '\0';
+    m_szDiscordMatchSecret[0] = '\0';
+    m_szDiscordSpectateSecret[0] = '\0';
+    m_szDiscordJoinSecret[0] = '\0';
+    m_sSteamLobbyID = CSteamID();
+    m_sSteamUserID = CSteamID();
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Set `m_sSteamUserID` once in the `PostInit()` call

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+    if (CSteamID(pParam->m_ulSteamIDLobby) != m_sSteamLobbyID) {
+        m_sSteamLobbyID = CSteamID(pParam->m_ulSteamIDLobby);
+        UpdateDiscordPartyIdFromSteam();
+    }
+
+    OnSteamLobbyUpdate();
+}
+
+void CMomentumDiscord::HandleLobbyChatUpdate(LobbyChatUpdate_t* pParam) {
+    if (CSteamID(pParam->m_ulSteamIDLobby) != m_sSteamLobbyID) {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Compare `pParam->m_ulSteamIDLobby == m_sSteamLobbyID.ConvertToUInt64()` instead

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+
+    OnSteamLobbyUpdate();
+}
+
+// --------------------------- CGameEventListner ---------------------------- //
+
+// Called when there is a game event we are listening for
+// Subscribe to events in `PostInit`
+void CMomentumDiscord::FireGameEvent(IGameEvent* event) {
+    if (!Q_strcmp(event->GetName(), "lobby_leave")) {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Only listening for one event here, this stringcmp isn't needed.

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+
+// ----------------------------- Custom Methods ----------------------------- //
+
+// Returns true if the player is in a map, false otherwise
+bool CMomentumDiscord::InMap() {
+    return MapName() != nullptr;
+}
+
+// Gets the current user's steam ID
+void CMomentumDiscord::GetSteamUserID() {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Remove this function

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+    }
+    else {
+        return SteamMatchmaking()->GetLobbyMemberData(g_pMomentumDiscord->m_sSteamLobbyID, *steamID, LOBBY_DATA_MAP);
+    }
+}
+
+// Joins a lobby from the Steam Lobby ID
+// returns true if we had to change the lobby
+bool CMomentumDiscord::JoinSteamLobbyFromID(const char* lobbyID) {
+    if (g_pMomentumDiscord->m_sSteamLobbyID.IsValid() && !V_strcmp(lobbyID, CFmtStrN<DISCORD_MAX_BUFFER_SIZE>("%lld", g_pMomentumDiscord->m_sSteamLobbyID.ConvertToUint64()))) {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: This method is part of the class, no need to use `g_pMomentumDiscord`

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+        m_szDiscordPartyId[0] = '\0';
+        m_szDiscordJoinSecret[0] = '\0';
+        m_szDiscordSpectateSecret[0] = '\0';
+    }
+}
+
+// ----------------------------- Custom Methods ----------------------------- //
+
+// Returns true if the player is in a map, false otherwise
+bool CMomentumDiscord::InMap() {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Change this to be a member variable boolean that gets set true on map load, false on level shutdown

File: mp/src/game/client/momentum/mom_system_discord.h [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+// (some things are still updated each frame such as checking callbacks)
+#ifndef DISCORD_FRAME_UPDATE_FREQ
+    #define DISCORD_FRAME_UPDATE_FREQ 600
+#endif
+
+// A class to manage the Discord Rich Presence feature
+// Uses the discord-rpc library
+// Libary source code is here: https://github.com/discordapp/discord-rpc
+// Docs are here: https://discordapp.com/developers/docs/rich-presence/how-to
+class CMomentumDiscord : public CAutoGameSystemPerFrame, CGameEventListener {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.h]: `public CGameEventListener`

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+#include <time.h>
+
+#include "tier0/memdbgon.h"
+
+// MOM_TODO: Change this to the official Discord App ID (currently it is under Connor's test account)
+const char* CMomentumDiscord::s_pDiscordAppID = "452297955718987776";
+const char* CMomentumDiscord::s_pSteamAppID = "669270";
+
+const char* CMomentumDiscord::s_pszInMenusStatusString = "In Menus";
+const char* CMomentumDiscord::s_pszInMenusLargeImage = "mom";
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: These should also be `#define`s

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    }
+    else {
+        m_sSteamLobbyID.Clear();
+        m_szDiscordPartyId[0] = '\0';
+        m_szDiscordJoinSecret[0] = '\0';
+    }
+}
+
+// Clears out all of the discord fields (optionally can preserve the party ID fields)
+void CMomentumDiscord::ClearDiscordFields(bool clearPartyFields /*=false*/) {
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Minor, it's actually defaulted to true here, but perhaps it should be a separate `Clear` method?

File: mp/src/game/client/momentum/mom_system_discord.cpp [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+            DevWarning("Error in lobby or user ID\n");
+            return;
+        }
+
+        // Try to get our target's current map
+        g_pMomentumDiscord->m_ulSpectateTargetUser = V_atoui64(g_pMomentumDiscord->m_szSpectateTargetUser);
+        CSteamID targetPlayerSteamID = CSteamID(g_pMomentumDiscord->m_ulSpectateTargetUser);
+        const char* targetMapName = GetMapOfPlayerFromSteamID(&targetPlayerSteamID);
+
+        bool inSameLobby = g_pMomentumDiscord->m_sSteamLobbyID.ConvertToUint64() == V_atoui64(g_pMomentumDiscord->m_szSpectateTargetLobby);
```
Comment (Gocnak) [inline mp/src/game/client/momentum/mom_system_discord.cpp]: Don't convert the string twice in a row. Use the new `m_ulSpectateTargetUser`

**Comment by SlidyBat** [comment]
Hi @connorjan, not sure if you're still active but I noticed that you have another branch in your fork in which you commited the changes that Gocnak suggested as well as some other fixes.

I'm planning on checking out that branch and testing it out (and possibly applying a couple of tweaks) so that we can get this merged.

Thanks for your contribution 😃.

**Comment by Gocnak** [comment]
Closing in favor of #168

**Pull request closed without merging**